### PR TITLE
Dont use a transaction in the scanner while using db based locking

### LINKED ISSUE #19457: 1205 Lock wait timeout exceeded when trying to access a file on external storage before it's ready
### Steps to reproduce

1. Add this code somewhere
```
$scanner = new \OC\Files\Utils\Scanner($user, \OC::$server->getDatabaseConnection());
$scanner->listen('\OC\Files\Utils\Scanner', 'scanFile', function ($path) use ($output) {
	$path = normalizePathToMakeItWorkWithPreview($path);
	$preview = new Preview($user, 'files', $path);
	$preview->setMaxX($this->thumbnailWidth);
	$preview->setMaxY($this->thumbnailHeight);
	$preview->setKeepAspect(true);
	$preview->getPreview();
});
```

1. Connect your instance to another oC instance using files_ external -> WebDAV
2. Upload a new file in the external cloud
3. Launch your script

Now the scanner is going loopy. It tries a hundred time to access the newly uploaded file, but it's never ready (unlocked?). A possible race condition? 

1. Change mount option to "Every time the filesytem is used" (value 2).
2. Launch your script again

This time it's worse, the scanner keeps trying to access the external folder until the command is killed.

### Notes

* I think this happens because `getPreview()` tries to access the file before it's stored in the cache
* New thumbnails, as they're generated, are also sent to the preview generator via the `scanFile` event
* When using `postScanFile` things seem to work as expected, but some people have reported experiencing the same issue with it
* Utils\Scanner **does not** use transactional locking

### Logs 

This is for the last operation, when always getting the folder.

```
{"reqId":"o2g1Y985c54xz\/5dY2C1","remoteAddr":"10.0.2.2","app":"webdav","message":"Exception: {\"Message\":\"An exception occurred while executing 'UPDATE `oc_file_locks` SET `lock` = `lock` + 1, `ttl` = ? WHERE `key` = ? AND `lock` >= 0' with params [1443562215, \\\"files\\\\\\\/a3594fd4eca7119cb937772e1f8f5415\\\"]:\\n\\nSQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction\",\"Exception\":\"Doctrine\\\\DBAL\\\\Exception\\\\DriverException\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/DBALException.php(116): Doctrine\\\\DBAL\\\\Driver\\\\AbstractMySQLDriver->convertException('An exception oc...', Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException))\\n#1 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Connection.php(996): Doctrine\\\\DBAL\\\\DBALException::driverExceptionDuringQuery(Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOMySql\\\\Driver), Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException), 'UPDATE `oc_file...', Array)\\n#2 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/db\\\/connection.php(203): Doctrine\\\\DBAL\\\\Connection->executeUpdate('UPDATE `oc_file...', Array, Array)\\n#3 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/lock\\\/dblockingprovider.php(112): OC\\\\DB\\\\Connection->executeUpdate('UPDATE `*PREFIX...', Array)\\n#4 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/common.php(638): OC\\\\Lock\\\\DBLockingProvider->acquireLock('files\\\/a3594fd4e...', 1)\\n#5 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/wrapper.php(571): OC\\\\Files\\\\Storage\\\\Common->acquireLock('files', 1, Object(OC\\\\Lock\\\\DBLockingProvider))\\n#6 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(1779): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->acquireLock('files', 1, Object(OC\\\\Lock\\\\DBLockingProvider))\\n#7 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(1874): OC\\\\Files\\\\View->lockPath('', 1, false)\\n#8 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(1190): OC\\\\Files\\\\View->lockFile('', 1)\\n#9 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/connector\\\/sabre\\\/serverfactory.php(83): OC\\\\Files\\\\View->getFileInfo('')\\n#10 [internal function]: OC\\\\Connector\\\\Sabre\\\\ServerFactory->OC\\\\Connector\\\\Sabre\\\\{closure}(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#11 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Object(Closure), Array)\\n#12 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#13 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#14 \\\/var\\\/www\\\/owncloud\\\/apps\\\/files\\\/appinfo\\\/remote.php(56): Sabre\\\\DAV\\\\Server->exec()\\n#15 \\\/var\\\/www\\\/owncloud\\\/remote.php(137): require_once('\\\/var\\\/www\\\/ownclo...')\\n#16 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Driver\\\/AbstractMySQLDriver.php\",\"Line\":115}","level":4,"time":"2015-09-29T20:31:06+00:00","method":"PROPFIND","url":"\/owncloud\/remote.php\/webdav\/"}
{"reqId":"Qb5n+Cq7CxO6bRP1sZRI","remoteAddr":"10.0.2.2","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 401 No basic authentication headers were found\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\NotAuthenticated\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/connector\\\/sabre\\\/auth.php(149): Sabre\\\\DAV\\\\Auth\\\\Backend\\\\AbstractBasic->authenticate(Object(OC\\\\Connector\\\\Sabre\\\\Server), 'ownCloud')\\n#1 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/connector\\\/sabre\\\/auth.php(122): OC\\\\Connector\\\\Sabre\\\\Auth->auth(Object(OC\\\\Connector\\\\Sabre\\\\Server), 'ownCloud')\\n#2 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Plugin.php(118): OC\\\\Connector\\\\Sabre\\\\Auth->authenticate(Object(OC\\\\Connector\\\\Sabre\\\\Server), 'ownCloud')\\n#3 [internal function]: Sabre\\\\DAV\\\\Auth\\\\Plugin->beforeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#5 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(456): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#6 \\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#7 \\\/var\\\/www\\\/owncloud\\\/apps\\\/files\\\/appinfo\\\/remote.php(56): Sabre\\\\DAV\\\\Server->exec()\\n#8 \\\/var\\\/www\\\/owncloud\\\/remote.php(137): require_once('\\\/var\\\/www\\\/ownclo...')\\n#9 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Backend\\\/AbstractBasic.php\",\"Line\":70}","level":0,"time":"2015-09-29T20:31:19+00:00","method":"PROPFIND","url":"\/owncloud\/remote.php\/webdav\/"}
```

### Config

```
$CONFIG = array (
  'datadirectory' => '/var/www/owncloud/data',
  'dbtype' => 'mysql',
  'version' => '8.2.0.6',
  'logtimezone' => 'UTC',
  'installed' => true,
  'loglevel' => 0,
  'theme' => '',
  'maintenance' => false,
  'enabledPreviewProviders' =>
  array (
    0 => 'OC\\Preview\\JPEG',
    1 => 'OC\\Preview\\PNG',
    2 => 'OC\\Preview\\GIF',
    3 => 'OC\\Preview\\Postscript',
    4 => 'OC\\Preview\\Font',
    10 => 'OC\\Preview\\Photoshop',
    11 => 'OC\\Preview\\Illustrator',
    12 => 'OC\\Preview\\TIFF',
    14 => 'OC\\Preview\\Raw',
  ),
  'preview_max_x' => 2048,
  'preview_max_y' => 2048,
  'debug' => true,
  'updatechecker' => 'false',
  'singleuser' => false,
  'memcache.local' => '\\OC\\Memcache\\APCu',
);
```

----

@icewind1991 @PVince81

### PR DESCRIPTION
**Pull request by icewind1991**
Fixes #19457

Scanning is the only place where we have transactions that commonly last longer then the locking timeout of the database which can cause issues when using the database locking provider.

cc @oparoz @DeepDiver1975 @PVince81

### CODE REVIEW

File: lib/private/files/utils/scanner.php [hunk 1]
```diff
@@ -156,9 +157,14 @@ public function scan($dir = '') {
 			$scanner = $storage->getScanner();
 			$scanner->setUseTransactions(false);
```
Comment (PVince81) [inline lib/private/files/utils/scanner.php]: what about these transactions ?

so here you're just disabling the "main" transaction ?

Comment (icewind1991) [inline lib/private/files/utils/scanner.php]: Per-folder transactions should still be disabled, otherwise it might still cause issues with large folders

**Comment by PVince81** [comment]
The code looks fine :+1: 
but I can't evaluate the potential regressions

**Comment by DeepDiver1975** [comment]
we need to smahbox this

@nickvergessen :speak_no_evil: THX

**Comment by nickvergessen** [comment]
> we need to smahbox this

running

**Comment by nickvergessen** [comment]
Passes :+1:

**Comment by MorrisJobke** [comment]
> Passes :+1: , but didn't try accessing a file on external storage

I tested with an external storage and it is accessible :+1:

**Comment by oparoz** [comment]
By accessible, you mean you were able to browse files while the scan was running?

**Comment by MorrisJobke** [comment]
> @MorrisJobke - By accessible, you mean you were able to browse files while the scan was running?

No. It started the scan and den showed the files. As usual.

**Comment by oparoz** [comment]
I've added this to the Gallery occ command and users can now use the GUI while a scan is taking place :+1:

**Pull request merged**